### PR TITLE
feat(provider/oraclebmcs): Instance cache agent, provider and tests

### PR DESCRIPTION
**Pull request by simonlord**
Oracle BMCS Instance cache agent, provider and tests.

### CODE REVIEW

File: clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/cache/Keys.groovy [hunk 1]
```diff
@@ -103,6 +112,13 @@ class Keys {
     "$OracleBMCSCloudProvider.ID:${Namespace.IMAGES}:${account}:${region}:${imageId}"
   }
 
+  static String getInstanceKey(String account,
+                               String region,
+                               String name,
+                               String id) {
+    "$OracleBMCSCloudProvider.ID:${Namespace.INSTANCES}:${account}:${region}:${name}:${id}"
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/cache/Keys.groovy]: Is the misalignment between this and `Keys` intentional?
```
account == account
region == region
name == instanceId
id == instanceOcid
```

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    creds.name = "foo"
+    creds.region = Region.US_PHOENIX_1.regionId
+    def computeClient = Mock(ComputeClient)
+
+    def instances = [
+      new Instance("AD1", null, "Instance 1", "ocid.instance.1", "ocid.image.1", null, Instance.LifecycleState.Running, null, creds.region, "small", null),
+      new Instance("AD1", null, "Instance 2", "ocid.instance.2", "ocid.image.1", null, Instance.LifecycleState.Starting, null, creds.region, "small", null),
+      new Instance("AD1", null, "Instance 3", "ocid.instance.3", "ocid.image.1", null, Instance.LifecycleState.Terminated, null, creds.region, "small", null),
+    ]
+
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy]: Extra blank line.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    then:
+    cacheResult != null
+    cacheResult.cacheResults.containsKey(INSTANCES.ns)
+    cacheResult.cacheResults.get(INSTANCES.ns).size() == 2
+    cacheResult.cacheResults.get(INSTANCES.ns).first().id == Keys.getInstanceKey(creds.name, creds.region, "Instance 1", "ocid.instance.1")
+    cacheResult.cacheResults.get(INSTANCES.ns).first().attributes.get("id") == "ocid.instance.1"
+    cacheResult.cacheResults.get(INSTANCES.ns).first().attributes.get("displayName") == "Instance 1"
+  }
+
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy]: Extra blank lines.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.cats.cache.DefaultCacheData
+import com.netflix.spinnaker.cats.mem.InMemoryCache
+import com.netflix.spinnaker.clouddriver.model.HealthState
+import com.netflix.spinnaker.clouddriver.oraclebmcs.cache.Keys
+import com.oracle.bmc.core.model.Instance
+import spock.lang.Specification
+
+class OracleBMCSInstanceProviderSpec extends Specification {
+
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy]: Extra blank line.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    def attributes = new ObjectMapper().convertValue(instance, new TypeReference<Map<String, Object>>() {})
+
+    return new DefaultCacheData(
+      Keys.getInstanceKey(account, region, name, ocid),
+      attributes as Map<String, Object>,
+      [:]
+    )
+  }
+
```
Comment (duftler) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy]: Extra blank line.

File: clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/cache/Keys.groovy (hunk 1 above)
Comment (duftler) [inline clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/cache/Keys.groovy]: Looks like it is, based on `OracleBMCSInstanceProvider.getInstance()`.

Comment (simonlord) [inline clouddriver-oracle-bmcs/src/main/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/cache/Keys.groovy]: Woops, no it's not intentional - i've updated the switch statement on line 60/61

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy (hunk 2 above)
Comment (simonlord) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy]: Updated so it's single blank line.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy (hunk 3 above)
Comment (simonlord) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/agent/OracleBMCSInstanceCachingAgentSpec.groovy]: Updated so it's single blank line.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy (hunk 4 above)
Comment (simonlord) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy]: Updated so it's single blank line.

File: clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy (hunk 5 above)
Comment (simonlord) [inline clouddriver-oracle-bmcs/src/test/groovy/com/netflix/spinnaker/clouddriver/oraclebmcs/provider/view/OracleBMCSInstanceProviderSpec.groovy]: Updated so it's single blank line.

**Pull request merged**